### PR TITLE
[Config] Allow scalar configuration in PHP Configuration

### PR DESCRIPTION
**Pull request by HypeMC**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix https://github.com/symfony/monolog-bundle/pull/417#issuecomment-1123294337
| License       | MIT
| Doc PR        | -

Fixes passing scalar values to array nodes that have a `beforeNormalization` hook, eg:

```php
return static function (MonologConfig $config): void {
    $config->handler('console')
        // ...
        ->processPsr3Messages()
            ->enabled(true)
    ;
};
```

can be shortened thanks to a [`beforeNormalization` hook](https://github.com/symfony/monolog-bundle/blob/a41bbcdc1105603b6d73a7d9a43a3788f8e0fb7d/DependencyInjection/Configuration.php#L453):

```php
return static function (MonologConfig $config): void {
    $config->handler('console')
        // ...
        ->processPsr3Messages(true)
    ;
};
```

I've used some of the code from https://github.com/symfony/symfony/pull/44166 by @jderusse. Since his PR is a feature it can't go on 5.4, but it still helped a lot.

### CODE REVIEW

File: src/Symfony/Component/Config/Builder/ConfigBuilderGenerator.php [hunk 1]
```diff
@@ -375,16 +393,16 @@ private function buildToArray(ClassBuilder $class): void
             $code = '$this->PROPERTY';
             if (null !== $p->getType()) {
                 if ($p->isArray()) {
-                    $code = 'array_map(function ($v) { return $v->toArray(); }, $this->PROPERTY)';
+                    $code = 'array_map(function ($v) { return $v instanceof CLASS ? $v->toArray() : $v; }, $this->PROPERTY)';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ConfigBuilderGenerator.php]: can we replace this by `return \is_array($v) ? $v : $v->toArray();` everywhere?

File: src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes/Symfony/Config/ScalarNormalizedTypes/ListObjectConfig.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace Symfony\Config\ScalarNormalizedTypes;
+
+
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes/Symfony/Config/ScalarNormalizedTypes/ListObjectConfig.php]: blank line :) (same after `use`)

**Review by nicolas-grekas — COMMENTED** [review]
Is there a way to do this only when the node actually has a beforeNormalization? Unless I missed something, right now this will always generate the adaptative code while it might not be needed when beforeNormalization isn't defined.

File: src/Symfony/Component/Config/Builder/ConfigBuilderGenerator.php (hunk 1 above)
Comment (HypeMC) [reply src/Symfony/Component/Config/Builder/ConfigBuilderGenerator.php]: @nicolas-grekas Actually no, `$this->PROPERTY` in an array of either `CLASS` objects or scalars, eg:

```php
return static function (ScalarNormalizedTypesConfig $config) {
    $config
        ->listObject('baz')
        ->listObject()->name('qux');
};
```

results in:

```
Symfony\Config\ScalarNormalizedTypesConfig {
  listObject: array:2 [
    0 => "baz"
    1 => Symfony\Config\ScalarNormalizedTypes\ListObjectConfig {
      name: "qux"
      data: null
      _usedProperties: array:1 [
        "name" => true
      ]
    }
  ]
  _usedProperties: array:1 [
    "listObject" => true
  ]
}
```

File: src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes/Symfony/Config/ScalarNormalizedTypes/ListObjectConfig.php (hunk 2 above)
Comment (HypeMC) [reply src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes/Symfony/Config/ScalarNormalizedTypes/ListObjectConfig.php]: @nicolas-grekas This file is auto-generated, I've adjusted the template https://github.com/symfony/symfony/pull/46328/commits/66a896916496037f350bcb550670d79e184d9592

**Comment by HypeMC** [comment]
> Is there a way to do this only when the node actually has a beforeNormalization? Unless I missed something, right now this will always generate the adaptative code while it might not be needed when beforeNormalization isn't defined.

@nicolas-grekas As far as I can tell there's no way to currently know if there are any `$normalizationClosures` set or not, a `hasNormalizationClosures` method could be added to the `BaseNode`, but that would be a new feature if I'm not mistaking.
Another way would be by using the reflection api.

**Comment by nicolas-grekas** [comment]
Let's do it by reflection then: that's within the same package, that's fine.

**Comment by HypeMC** [comment]
> Let's do it by reflection then: that's within the same package, that's fine.

@nicolas-grekas The adaptive code is now conditionally generated depending on whether there are normalization closures set or not.  I added a new method to the `Builder/Property` class, I think this shouldn't be a problem since it's an internal class, but even if it is a problem, the code can easily be moved to a private method in `ConfigBuilderGenerator`, it's just a little cleaner this way.

File: src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
+<?php
+
+namespace Symfony\Component\Config\Tests\Builder\Fixtures;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes.php:12]: please add the licence header to non-generated fixtures

**Review by nicolas-grekas — APPROVED** [review]
Good one, thank you!

File: src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes.php (hunk 3 above)
Comment (HypeMC) [reply src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes.php:12]: @nicolas-grekas Done, https://github.com/symfony/symfony/pull/46328/commits/9d16892ce773d4a26603245e13e6f6165541f181

**Comment by nicolas-grekas** [comment]
Thank you @HypeMC.

**Comment by nicolas-grekas** [comment]
And thank you @jderusse also!

**Comment by nicolas-grekas** [comment]
I've merged this up into 6.0 but there is a failing test:

>1) Symfony\Component\Config\Tests\Builder\GeneratedConfigTest::testConfig with data set "ScalarNormalizedTypes" ('ScalarNormalizedTypes', 'scalar_normalized_types')
TypeError: Symfony\Config\ScalarNormalizedTypesConfig::simpleArray(): Argument `#`1 ($value) must be of type Symfony\Component\Config\Loader\ParamConfigurator|array, string given, called in src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes.config.php on line 16

Could you please have a look and send a fix? :pray:

**Comment by nicolas-grekas** [comment]
I'd be also happy to let you merge 6.0 into 6.1 if you're up to :grimacing: (you can do it on your fork and let me know the commit hash here.)

**Comment by HypeMC** [comment]
@nicolas-grekas Sure, I'll take a look.

**Comment by HypeMC** [comment]
> I'd be also happy to let you merge 6.0 into 6.1 if you're up to grimacing (you can do it on your fork and let me know the commit hash here.)

@nicolas-grekas Done https://github.com/HypeMC/symfony/commit/123b1651c4a7e219ba59074441badfac65525efe

**Comment by nicolas-grekas** [comment]
Great thank you so much! Now into 6.1.

**Pull request merged**